pvanalytics/features/clipping.py

```diff
@@ -88,3 +88,141 @@ def levels(ac_power, window=4, fraction_in_window=0.75,
+        freq = pd.Timedelta(pd.infer_freq(ac_power.index)).seconds / 60
+    elif isinstance(freq, str):
+        freq = pd.Timedelta(freq).seconds / 60
+
+    # Use the derivative of the 99.5% quantile of daytime power at
+    # each minute to identify clipping.
+    powercurve = _daytime_powercurve(ac_power)
+    normalized_power = powercurve / powercurve.max()
+    power_derivative = (normalized_power.diff()
+                        / normalized_power.index.to_series().diff()) * freq
+
+    clipped_times = _clipped(powercurve, power_derivative,
+                             powercurve.median() * power_min,
+                             derivative_max)
+    clipping_cumsum = (~clipped_times).cumsum()
+    # get the value of the cumulative sum the longest True span
+    longest_clipped = clipping_cumsum.value_counts().idxmax()
+    # select the longest span that satisfies the clipping criteria
+    longest = powercurve[clipping_cumsum == longest_clipped]
+
+    if longest.index[-1] - longest.index[0] >= 60:
+        # if the period of clipping is at least 60 minutes then we
+        # have enough evidence to determine the clipping threshold.
+        return longest.mean()
+    return None
+
+
+def threshold(ac_power, derivative_max=0.0035,
+              power_quantile=0.995, frequency_quantile=0.25,
+              freq=None):
+    """Detect clipping based on a maximum power threshold.
+
+    A clipping threshold is calculated from the AC power data and any
+    power greater than the threshold is flagged as clipped.
+
+    To calculate the clipping threshold, `ac_power` is aggregated at
+    each minute of the day. Low power data is removed to eliminate
+    night-time periods and the 99.5% quantile is computed at each
+    minute. If the absolute value of the derivative of the 99.5%
+    quantile (with respect to time) is less than `derivative_max` for
+    a continuous period of at least one hour then clipping is
+    indicated. The mean power for that period is used as the
+    threshold. If there are multiple periods with a derivative less
+    than `derivative_max` then the longest period is used to compute
+    the threshold.
+
+    Parameters
+    ----------
+    ac_power : Series
+        DatetimeIndexed series of AC power data.
+    derivative_max : float, default 0.0035
+        Maximum absolute value of derivative of AC power quantile for
+        clipping to be indicated. The default value has been derived
+        empirically to prevent false positives for tracking PV
+        systems.
+    power_quantile : float, default 0.995
+        quantile used to determine the maximum power for each minute
+        of the day
+    frequency_quantile : float, default 0.25
+        After taking the count of positive values in `ac_power` at
+        each minute of the day, any minute with a count less than the
+        `frequency_quantile`-quantile of all counts is excluded from
+        the calculation of the clipping threshold.
```

Comment:
Is this a reference to the day/ night detector?

Reply:
I'm glad you brought this up. We do a sort of day/night filtering step in `_daytime_powercurve`. It might be better to pass in only daytime power, or pass a day/night mask. Passing a mask is probably the best solution to limit coupling between modules while still reducing code duplication.

@matt14muller, are there any caveats/pitfalls you can see in changing the original algorithm like this?

Reply:
I'm actually starting to wonder if this step is even necessary. We only accept a period with low slope a clipping if it has sufficiently high power (see the `power_min` parameter). That should prevent using night time or early morning periods where we shouldn't be seeing clipping as the the clipping threshold.

Reply:
I am not sure I follow your question.   Can you elaborate?

Reply:
There is a step where minutes of the day with a low number of positive values are eliminated. I think it is to remove the night/early morning/evening times. Do you think that is really necessary? Can we just include night-time/morning/evening when looking for clipping?

Reply:
>  Can we just include night-time/morning/evening when looking for clipping?

We can with this algorithm, since there is a power threshold critieria. The SFA function only looks for flat spots, and has to be combined with a night/day detector to avoid labeling night as clipping.

Reply:
I'll need to remove the `frequency_quantile` parameter (and associated code) and see if the tests still pass.

Reply:
@cwhanse @matt14muller I removed the `frequency_quantile` param. Works fine for the tests, but those are just simple tests... The only problem I can think of with this is if there is enough erroneous high data in the night time that the 95% quantile is driven up. I'm still good with eliminating the parameter though since we have many other quality functions to help users eliminate that problem. If worst comes to worst, users can simply pass only daytime power values, leaving night time out of the equation all together.